elastic-container.sh

```diff
@@ -15,92 +15,185 @@
--e "KIBANA_FLEET_SETUP=1" \
--e "FLEET_SERVER_ENABLE=1" \
--e "FLEET_SERVER_INSECURE_HTTP=1" \
-docker.elastic.co/beats/elastic-agent:${STACK_VERSION}
-echo
-echo "Browse to http://localhost:5601"
-echo "Username: elastic"
-echo "Passphrase: ${ELASTIC_PASSWORD}"
-
-# Stop and remove the Elastic containers and network
-else
-if [ $1 == stop ] 2> /dev/null
-then
-echo "#####"
-echo "Stopping and removing all Elastic Stack components."
-echo "#####"
-docker stop fleet-server 2> /dev/null
-docker stop kibana 2> /dev/null
-docker stop elasticsearch 2> /dev/null
-docker network rm elastic 2> /dev/null
-
-# Restart the Elastic containers
-else
-if [ $1 == restart ] 2> /dev/null
-then
-echo "#####"
-echo "Restarting all Elastic Stack components."
-echo "#####"
-docker restart elasticsearch 2> /dev/null
-docker restart kibana 2> /dev/null
-docker restart fleet-server 2> /dev/null
-
-# Get the status of the Elastic containers
-else
-if [ $1 == status ] 2> /dev/null
-then
-docker ps -f "name=kibana" -f "name=elasticsearch" -f "name=fleet-server" --format "table {{.Names}}: {{.Status}}"
+  actions:
+    stage     downloads all necessary images to local storage
+    start     creates network and configures containers to run
+    stop      stops the containers created and removes the network
+    restart   simply restarts all the stack containers
+    status    check the status of the stack containers
+    help      print this message
+
+  flags:
+    -v        enable verbose output
+EOF
+}
+
+configure_kbn() {
+  MAXTRIES=15
+  i=${MAXTRIES}
+
+  while [ $i -gt 0 ]; do
+    STATUS=$(curl -I "${LOCAL_KBN_URL}" 2>&3 | head -n 1 | cut -d$' ' -f2)
+
+    echo
+    echo "Attempting to enable the Detection Engine and Prebuilt-Detection Rules"
+
+    if [ "${STATUS}" == "302" ]; then
+      echo
+      echo "Kibana is up. Proceeding"
+      output=$(curl --silent "${HEADERS[@]}" --user "${ELASTIC_USERNAME}:${ELASTIC_PASSWORD}" -XPOST "${LOCAL_KBN_URL}/api/detection_engine/index" 2>&3)
+      [[ $output =~ '"acknowledged":true' ]] || (
+        echo "Detection Engine setup failed :-("
+        exit 1
+      )
+
+      echo "Detection engine enabled. Installing prepackaged rules."
+      curl --silent "${HEADERS[@]}" --user "${ELASTIC_USERNAME}:${ELASTIC_PASSWORD}" -XPUT "${LOCAL_KBN_URL}/api/detection_engine/rules/prepackaged" 1>&3 2>&3
+
+      echo
+      echo "Prebuilt Detections Enabled!"
+      break
+    else
+      echo
+      echo "Kibana still loading. Trying again in 40 seconds"
+    fi
+
+    sleep 40
+    i=$((i - 1))
+  done
+
+  [ $i -eq 0 ] && echo "Exceeded MAXTRIES (${MAXTRIES}) to setup detection engine." && exit 1
+}
+
+# A POSIX variable
+OPTIND=1 # Reset in case getopts has been used previously in the shell.
+
+# Initialize our own variables:
+verbose=0
+
+while getopts "v" opt; do
+  case "$opt" in
+  v)
+    verbose=1
+    ;;
+  *) ;;
+  esac
+done
+
+shift $((OPTIND - 1))
+
+[ "${1:-}" = "--" ] && shift
+
```

Comment:
What's this?

Reply:
One word @dcode

Reply:
Pretty positive it adds verbose output to the options with -v though